natchez-http4s/src/main/scala/com/ovoenergy/effect/natchez/http4s/server/Configuration.scala

```diff
@@ -0,0 +1,169 @@
+package com.ovoenergy.effect.natchez.http4s.server
+
+import cats.Applicative
+import cats.data.Kleisli
+import cats.effect.Sync
+import cats.instances.list._
+import cats.instances.map._
+import cats.kernel.{Monoid, Semigroup}
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.monoid._
+import com.ovoenergy.effect.natchez.http4s.server.Configuration.TagReader._
+import natchez.TraceValue
+import natchez.TraceValue.StringValue
+import org.http4s.util.{CaseInsensitiveString, StringWriter}
+import org.http4s.{Headers, Message, Request, Response}
+
+/**
+ * The tricky part about putting HTTP4s middleware into a library is that
+ * each user is likely to want to extract different tags from their requests
+ * according to whatever their organisation wide tagging policy is
+ *
+ * As such we define how to create tags from Requests / Responses with a `TagReader`
+ * which has a semigroup instance, allowing you to cherry pick what you want to extract
+ * and write your own extractors if required.
+ */
+case class Configuration[F[_]](
+  request: RequestReader[F],
+  response: ResponseReader[F]
+)
+
+object Configuration {
+
+  // just to shorten some lines
+  type Tags = Map[String, TraceValue]
+
+  /**
+   * A tag reader is essentially a function from an HTTP message to F[Tags]
+   * We need the effect because some tags may involve streaming the entity body
+   */
+  case class TagReader[F[_], -A](value: Kleisli[F, A, Tags]) extends AnyVal
+
+  object TagReader {
+
+    type MessageReader[F[_]] = TagReader[F, Message[F]]
+    type RequestReader[F[_]] = TagReader[F, Request[F]]
+    type ResponseReader[F[_]] = TagReader[F, Response[F]]
+
+    /**
+     * Monoid instance for TagReader
+     * so it is easy to read many tags from a request
+     */
+    implicit def monoid[F[_]: Applicative, A]: Monoid[TagReader[F, A]] = {
+      implicit def underlying[B: Monoid]: Monoid[F[B]] = Applicative.monoid[F, B]
+      implicit val takeLast: Semigroup[TraceValue] = (_, b) => b
+
+      new Monoid[TagReader[F, A]] {
+        def empty: TagReader[F, A] =
+          TagReader(Monoid[Kleisli[F, A, Map[String, TraceValue]]].empty)
+        def combine(x: TagReader[F, A], y: TagReader[F, A]): TagReader[F, A] =
+          TagReader(x.value |+| y.value)
+      }
+    }
+
+    def message[F[_]: Applicative](f: Message[F] => Tags): MessageReader[F] =
+      TagReader(Kleisli(a => Applicative[F].pure(f(a))))
+
+    def request[F[_]: Applicative](f: Request[F] => Tags): RequestReader[F] =
+      TagReader(Kleisli(a => Applicative[F].pure(f(a))))
+
+    def response[F[_]: Applicative](f: Response[F] => Tags): ResponseReader[F] =
+      TagReader(Kleisli(a => Applicative[F].pure(f(a))))
+  }
+
+  private val isSensitive: CaseInsensitiveString => Boolean =
+    cs => Headers.SensitiveHeaders.contains(cs) || cs.value.toLowerCase.contains("key")
```

Comment:
Would it make sense to make this configurable? I guess what headers are considered sensitive will be specific to each service, so it's worth allowing users to provide their own set of headers here.

Reply:
:+1: I made it a param in 34adfb3

Reply:
thanks Tom